futures-util/src/task/arc_wake.rs

```diff
@@ -17,26 +17,41 @@ pub trait ArcWake {
     ///
     /// Executors generally maintain a queue of "ready" tasks; `wake` should place
     /// the associated task onto this queue.
-    fn wake(arc_self: &Arc<Self>);
+    fn wake(self: Arc<Self>);
+
+    /// Indicates that the associated task is ready to make progress and should
+    /// be `poll`ed.
+    ///
+    /// This function can be called from an arbitrary thread, including threads which
+    /// did not create the `ArcWake` based `Waker`.
+    ///
+    /// Executors generally maintain a queue of "ready" tasks; `wake_by_ref` should place
+    /// the associated task onto this queue.
+    ///
+    /// This function is similar to `wake`, but must not consume the provided data
+    /// pointer.
+    fn wake_by_ref(arc_self: &Arc<Self>) {
+        arc_self.clone().wake()
```

Comment:
I'm not certain whether it makes sense to keep the separation of waking by-value and by-ref for `ArcWake`, it looks like every implementation of `ArcWake::wake` delegates to `ArcWake::wake_by_ref`. But maybe the thread pool implementation can be updated to take advantage of it.

I think it makes more sense to provide the `ArcWake::wake` implementation as delegating to `ArcWake::wake_by_ref`, that would allow dropping all the current delegating implementations, and means that implementors wouldn't accidentally cause an unnecessary clone by only implementing `ArcWake::wake` (they may still cause an unnecessary clone by only implementing `ArcWake::wake_by_ref` and not taking advantage of being able to reuse the `Arc` in `ArcWake::wake`, but that seems like a more obvious performance issue since the clone will be happening in their code instead of the provided method).

Reply:
> I think it makes more sense to provide the `ArcWake::wake` implementation as delegating to `ArcWake::wake_by_ref`, that would allow dropping all the current delegating implementations, and means that implementors wouldn't accidentally cause an unnecessary clone by only implementing `ArcWake::wake` (they may still cause an unnecessary clone by only implementing `ArcWake::wake_by_ref` and not taking advantage of being able to reuse the `Arc` in `ArcWake::wake`, but that seems like a more obvious performance issue since the clone will be happening in their code instead of the provided method).

Thanks for pointing that out, that makes a lot of sense.